homeassistant/components/recorder/purge.py

```diff
@@ -619,102 +620,135 @@ def _purge_filtered_data(instance: Recorder, session: Session) -> bool:
     _LOGGER.debug("Cleanup filtered data")
     database_engine = instance.database_engine
     assert database_engine is not None
+    now_timestamp = time.time()
 
     # Check if excluded entity_ids are in database
     entity_filter = instance.entity_filter
+    has_more_states_to_purge = False
     excluded_metadata_ids: list[str] = [
```

Comment:
This is inefficient because every iteration is doing a full table scan of StatesMeta but it's still way more efficient that it was in 2023.3.x. We could refactor this to resolve them only the first time the service called and keep pushing the ids to the next task.  But that's for a future PR